<a href="https://colab.research.google.com/github/DanielRaiicHu/Pandas_Input_Output/blob/main/Pandas_Input_Output.ipynb?forceRefresh=true">
  <img src="https://drive.google.com/uc?export=view&id=191Woxwut0iaIL2hCcCIeXRSB6oba_88H" alt="Abrir en Colab" width="120"/>
</a>

# **Trabajando con distintos formatos de archivos**

## **Aula 1: Leyendo archivos csv**

Valores Separados por Comas (Comma Separated Values en inglés)

### **Importando archivos CSV** (Extraer/Leer)

Documentación: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

In [26]:
import pandas as pd

#### **Archivos separados por comas**

In [27]:
# Leyendo el archivo CSV con comas.
datos = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/superstore_data.csv')


In [28]:
# Para mostrar el DataFrame
datos.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

#### **Archivos separados por punto y coma**

In [29]:
# Leyendo el archivo CSV con punto y coma.
datos_punto_coma = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/superstore_data_punto_coma.csv',sep=';')

In [30]:
datos_punto_coma.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

#### **nrows (Lectura de filas)**
Número de filas del archivo a leer.

In [31]:
# Leyendo el archivo CSV con comas.
datos_primeras_lineas = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/superstore_data.csv', nrows = 5) #Leo solamente las primeras 5 filas
datos_primeras_lineas

Id  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced   84835        0         0   
1      1        1961  Graduation         Single   57091        0         0   
2  10476        1958  Graduation        Married   67267        0         1   
3   1386        1967  Graduation       Together   32474        1         1   
4   5371        1989  Graduation         Single   21474        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

#### **usecols (Lectura de columnas)**
Seleccionar las columnas que deseo ocupar.

In [32]:
# Colocamos los nombres de las columnas que necesitamos.
datos_seleccion = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/superstore_data.csv', usecols = ['Id','Year_Birth','Income'])
datos_seleccion

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [33]:
# Colocamos el índice de las columnas que necesitamos.
datos_seleccion_idx = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/superstore_data.csv', usecols = [0,1,4])
datos_seleccion_idx

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

### **Exportando archivos CSV** (Guardar/Escribir)

Documentación: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

#### **Guardando el archivo CSV**

In [34]:
# Guarda el DataFrame en el archivo
datos_seleccion.to_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/clientes_mercado.csv')

In [35]:
# Leemos el archivo creado
clientes_mercado = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/clientes_mercado.csv')
clientes_mercado
# Nos damos cuenta que generó una columna extra, con el índice.

Unnamed: 0     Id  Year_Birth   Income
0              0   1826        1970  84835.0
1              1      1        1961  57091.0
2              2  10476        1958  67267.0
3              3   1386        1967  32474.0
4              4   5371        1989  21474.0
...          ...    ...         ...      ...
2235        2235  10142        1976  66476.0
2236        2236   5263        1977  31056.0
2237        2237     22        1976  46310.0
2238        2238    528        1978  65819.0
2239        2239   4070        1969  94871.0

[2240 rows x 4 columns]

#### **Índice (index)**

Valor default es index=True

In [36]:
# Cambiamos el valor del índice a index=False
datos_seleccion.to_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/seleccion_mercado.csv', index=False)

In [37]:
# Leemos nuevamente nuestros datos y ahora sí están solo las columnas seleccionadas
clientes_mercado = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/seleccion_mercado.csv')
clientes_mercado

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

### Desafío: leyendo con CSV

Es hora de que pongas a prueba los conocimientos desarrollados durante la clase. Disponemos de un [archivo en formato CSV](https://raw.githubusercontent.com/alura-es-cursos/1980-Pandas_IO/datos_sus/datos_sus.csv), el cual se obtuvo de la página [Información de Salud (TABNET) - DATASUS](https://datasus.saude.gov.br/informacoes-de-saude-tabnet/). Este archivo contiene los gastos hospitalarios públicos de cada una de las Unidades de la Federación Brasileña desde enero de 2008 hasta marzo de 2021.



Su desafío aquí será leer este archivo usando la función read_csv de la biblioteca Pandas.

#### Ejercicio:

1. Asegúrese de que el archivo CSV esté separado por una coma o punto y coma.
2. La codificación del archivo es ISO-8859-1.
3. Las primeras tres líneas del archivo se pueden ignorar, ya que el encabezado sólo comienza en la cuarta línea.
4. Las últimas 9 líneas también se pueden ignorar, ya que son sólo información sobre dónde se tomaron los datos.
5. Para eliminar las últimas líneas es necesario agregar el parámetro engine='python'.

| Parámetro              | Descripción breve                                                                                   |
|------------------------|-----------------------------------------------------------------------------------------------------|
| `pd.read_csv()`        | Función de pandas que carga un archivo CSV en un DataFrame.                                         |
| `'ruta/del/archivo.csv'` | Ruta del archivo que estás leyendo desde Google Drive.                                              |
| `sep=';'`              | Indica que las columnas del archivo están separadas por punto y coma (`;`).                        |
| `encoding='ISO-8859-1'`| Codificación que permite leer caracteres especiales como tildes o eñes en español.                  |
| `skiprows=3`           | Omite las primeras 3 filas del archivo (pueden contener notas o encabezados innecesarios).         |
| `engine='python'`      | Usa el motor de lectura en Python (más flexible), **necesario para usar `skipfooter`**.             |
| `skipfooter=9`         | Ignora las últimas 9 filas del archivo (suelen ser notas, totales o datos fuera del formato útil).  |


In [38]:
datos_hospital = pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_csv/datos_sus.csv',sep=';', encoding='ISO-8859-1',skiprows=3,engine='python',skipfooter=9 )
datos_hospital



Unidade da Federação      2008/Jan      2008/Fev      2008/Mar  \
0              Rondônia    1388528,39    2931283,42    1541682,52   
1                  Acre     902416,00    1497206,26    1794028,48   
2              Amazonas    4735529,42    7118990,57    8196635,49   
3               Roraima     657889,53     777939,31     718688,03   
4                  Pará   18864744,11   19553758,20   21937342,70   
5                 Amapá    1120973,34    1268740,33     974651,24   
6             Tocantins    3441592,69    4985688,85    6205612,60   
7              Maranhão   10723506,39   13559679,41   13985142,36   
8                 Piauí    7985142,43   10079928,25   10298372,45   
9                 Ceará   15539821,90   25735101,55   28724625,75   
10  Rio Grande do Norte    5947814,08    9509755,83   10352238,59   
11              Paraíba   10723002,45   11386274,47   12053643,87   
12           Pernambuco   14393949,89   24181400,95   29991302,71   
13              Alagoas    6243726,32    8376622,16    8408336,22   
14              Sergipe    2780750,11    3676483,16    5051473,61   
15                Bahia   21211522,47   28189304,95   34181483,52   
16         Minas Gerais   45671659,76   70194037,10   79231336,95   
17       Espírito Santo    5456723,98    9102212,86   10249595,97   
18       Rio de Janeiro   23025258,52   28394310,93   44567062,94   
19            São Paulo   91678473,48  148677526,82  156087269,23   
20               Paraná   38266347,46   55384670,52   53801318,85   
21       Santa Catarina   16504913,45   25651429,11   26743524,89   
22    Rio Grande do Sul   30931325,46   57363420,01   51547794,07   
23   Mato Grosso do Sul    5377392,11    7002472,61    8423933,19   
24          Mato Grosso    4795299,01    8684882,23    7886243,14   
25                Goiás   15310637,29   22258668,68   21442515,75   
26     Distrito Federal    4745919,19    7632480,67    9647203,26   
27                Total  408424859,23  613174269,21  664043058,38   

        2008/Abr      2008/Mai      2008/Jun      2008/Jul      2008/Ago  \
0     1525314,96    1645953,84    1406150,68    3065279,01    3231494,61   
1     1730469,42    1819443,92    1828496,00    2511754,59    2089107,14   
2     8259378,42    7831399,90    8477273,62    9368858,72    9352532,70   
3      839994,39     862347,96     832440,66     996693,09     894271,18   
4    20842829,69   23249952,88   23240687,56   24002223,56   23341218,03   
5     1141137,32     866346,18    1440978,95    1426659,48    1475551,38   
6     4984482,09    5511002,46    4927934,22    5445602,28    5413710,38   
7    14809416,46   14505727,38   14908091,32   15784899,40   15195312,89   
8    10775976,33   10398360,88   10528253,88   11040107,92   10549176,40   
9    30676457,06   30604453,47   31272395,18   30925391,36   27851438,72   
10   10855108,50   11160168,19   10475071,38   11386349,96   10640817,47   
11   12653528,76   12215107,11   12256928,49   12755413,77   12872578,37   
12   34806197,97   32147987,35   32239094,89   32685640,76   35457038,66   
13    9090794,82    9338771,45   10068210,08    9887061,61    9373432,00   
14    5046063,36    4741151,56    5401259,63    5521904,44    5238398,50   
15   39026934,13   42949566,82   47025972,04   47481695,17   46931248,28   
16   77136537,58   80554440,63   94758522,83   85360064,11   84722528,10   
17   11253833,56   10198619,17   13364258,29   13035374,08   12588595,77   
18   44054604,82   44911021,15   48056212,24   49872379,35   50912706,60   
19  161227125,43  187119397,64  187291126,07  194398399,48  184854516,35   
20   52460306,69   54419620,71   57811688,48   58094241,83   56982774,90   
21   27181488,26   26845308,18   29909762,04   30789135,84   31250854,33   
22   56673338,49   48397733,07   52413429,62   56636376,98   56113054,70   
23    8797487,68    9770421,80   10905392,18   10249299,82    9671553,45   
24    8794109,79    9231645,40   10221763,77    9163314,28    9243148,81   
25   20921212,45   216159

## **Aula 2: Leyendo archivos Excel**

### **Importando archivos Excel**

In [39]:
import pandas as pd

In [40]:
# Crea una variable para almacenar la ubicación de nuestro archivo excel
archivo_excel = '/content/drive/MyDrive/Pandas_2/archivos_excel/emisiones_CO2.xlsx'

In [41]:
# Importando el archivo excel (En este caso, default importa la primera hoja de cálculo)
datos_co2 = pd.read_excel(archivo_excel)

In [42]:
# Muestra el DataFrame de la primera hoja
datos_co2.head(10)

País ISO 3166-1 alpha-3   Año  Total  Carbón  Aceite  Gas  Cemento  \
0  Afganistán                AFG  1750    0.0     NaN     NaN  NaN      NaN   
1  Afganistán                AFG  1751    0.0     NaN     NaN  NaN      NaN   
2  Afganistán                AFG  1752    0.0     NaN     NaN  NaN      NaN   
3  Afganistán                AFG  1753    0.0     NaN     NaN  NaN      NaN   
4  Afganistán                AFG  1754    0.0     NaN     NaN  NaN      NaN   
5  Afganistán                AFG  1755    0.0     NaN     NaN  NaN      NaN   
6  Afganistán                AFG  1756    0.0     NaN     NaN  NaN      NaN   
7  Afganistán                AFG  1757    0.0     NaN     NaN  NaN      NaN   
8  Afganistán                AFG  1758    0.0     NaN     NaN  NaN      NaN   
9  Afganistán                AFG  1759    0.0     NaN     NaN  NaN      NaN   

   Quema  Otros  Per Capita  
0    NaN    NaN         NaN  
1    NaN    NaN         NaN  
2    NaN    NaN         NaN  
3    NaN    NaN         NaN  
4    NaN    NaN         NaN  
5    NaN    NaN         NaN  
6    NaN    NaN         NaN  
7    NaN    NaN         NaN  
8    NaN    NaN         NaN  
9    NaN    NaN         NaN

In [43]:
# Muestra el nombre de las hojas de cálculo
pd.ExcelFile(archivo_excel).sheet_names

# En este ejemplo muestra 3 hojas de cálculo con sus respectivos nombres

['emisiones_C02', 'emisiones_percapita', 'fuentes']

### **Parámetros de la función read_excel**

Documentación: https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html

#### **sheet_name**

In [44]:
# Crea un DataFrame con los datos de la segunda hoja del archivo excel
# sheet_name -> el nombre de la hoja de cálculo
# archivo_excel -> la variable de la ubicación del archivo
percapita = pd.read_excel(archivo_excel, sheet_name='emisiones_percapita')

In [45]:
# Muestra las últimas 5 filas del DataFrame
percapita.tail()

País ISO 3166-1 alpha-3   Año      Total     Carbón     Aceite  \
63099  Global                WLD  2017  4749682.0  1908857.0  1610910.0   
63100  Global                WLD  2018  4792753.0  1919213.0  1596350.0   
63101  Global                WLD  2019  4775633.0  1896468.0  1589920.0   
63102  Global                WLD  2020  4497423.0  1807760.0  1427353.0   
63103  Global                WLD  2021  4693699.0  1893923.0  1496614.0   

             Gas   Cemento    Quema    Otros  
63099   940144.0  198416.0  51579.0  39776.0  
63100   979965.0  204225.0  53634.0  39366.0  
63101   984878.0  208309.0  56569.0  39490.0  
63102   963695.0  208844.0  51981.0  37789.0  
63103  1001585.0  211472.0  52663.0  37443.0

In [46]:
# Muestra (5) filas aleatorias del DataFrame
percapita.sample(5)

País ISO 3166-1 alpha-3   Año  Total  \
55390                           Taiwán                TWN  1924    NaN   
4135                           Bahrein                BHR  1805    NaN   
14431  República Democrática Del Congo                COD  1765    NaN   
36315                           México                MEX  1889    NaN   
15957                          Ecuador                ECU  1931    NaN   

       Carbón  Aceite  Gas  Cemento  Quema  Otros  
55390     NaN     NaN  NaN      NaN    NaN    NaN  
4135      NaN     NaN  NaN      NaN    NaN    NaN  
14431     NaN     NaN  NaN      NaN    NaN    NaN  
36315     NaN     NaN  NaN      NaN    NaN    NaN  
15957     NaN     NaN  NaN      NaN    NaN    NaN

In [47]:
fuentes = pd.read_excel(archivo_excel, sheet_name='fuentes')

In [48]:
fuentes.tail()

País ISO 3166-1 alpha-3   Año                                 Total  \
63099  Global                WLD  2017  CDIAC 2022, BP, and Sum of countries   
63100  Global                WLD  2018  CDIAC 2022, BP, and Sum of countries   
63101  Global                WLD  2019  CDIAC 2022, BP, and Sum of countries   
63102  Global                WLD  2020  CDIAC 2022, BP, and Sum of countries   
63103  Global                WLD  2021  CDIAC 2022, BP, and Sum of countries   

                  Carbón             Aceite                Gas        Cemento  \
63099  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63100  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63101  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63102  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63103  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   

                    Quema   Otros  \
63099  CDIAC 2022 and GCP  [NONE]   
63100  CDIAC 2022 and GCP  [NONE]   
63101  CDIAC 2022 and GCP  [NONE]   
63102  CDIAC 2022 and GCP  [NONE]   
63103  CDIAC 2022 and GCP  [NONE]   

                                              Per Capita  
63099  CDIAC 2022, BP, Sum of countries, and UN popul...  
63100  CDIAC 2022, BP, Sum of countries, and UN popul...  
63101  CDIAC 2022, BP, Sum of countries, and UN popul...  
63102  CDIAC 2022, BP, Sum of countries, and UN popul...  
63103  CDIAC 2022, BP, Sum of countries, and UN popul...

#### **use_cols (Selecciona un intervalo de columnas)**

In [49]:
# Para seleccionar un intervalo de columnas.
# En este caso seleccionamos las columnas desde la A y D.
intervalo = pd.read_excel(archivo_excel, sheet_name='emisiones_C02',usecols='A:D')

In [50]:
intervalo

País ISO 3166-1 alpha-3   Año         Total
0      Afganistán                AFG  1750  0.000000e+00
1      Afganistán                AFG  1751  0.000000e+00
2      Afganistán                AFG  1752  0.000000e+00
3      Afganistán                AFG  1753  0.000000e+00
4      Afganistán                AFG  1754  0.000000e+00
...           ...                ...   ...           ...
63099      Global                WLD  2017  3.609674e+10
63100      Global                WLD  2018  3.682651e+10
63101      Global                WLD  2019  3.708256e+10
63102      Global                WLD  2020  3.526409e+10
63103      Global                WLD  2021  3.712385e+10

[63104 rows x 4 columns]

#### **nrows (Selecciona una cantidad de filas)**

In [51]:
intervalo_2 = pd.read_excel(archivo_excel, sheet_name='emisiones_C02',usecols='A:D', nrows=10)

In [52]:
intervalo_2

País ISO 3166-1 alpha-3   Año  Total
0  Afganistán                AFG  1750      0
1  Afganistán                AFG  1751      0
2  Afganistán                AFG  1752      0
3  Afganistán                AFG  1753      0
4  Afganistán                AFG  1754      0
5  Afganistán                AFG  1755      0
6  Afganistán                AFG  1756      0
7  Afganistán                AFG  1757      0
8  Afganistán                AFG  1758      0
9  Afganistán                AFG  1759      0

### **Exportando archivos excel**

#### **Excel**

In [53]:
# Guarda el DataFrame en el archivo
intervalo.to_excel('/content/drive/MyDrive/Pandas_2/archivos_excel/co2_percapita.xlsx', index=False)

In [54]:
# Leemos el archivo creado
pd.read_excel('/content/drive/MyDrive/Pandas_2/archivos_excel/co2_percapita.xlsx')

País ISO 3166-1 alpha-3   Año         Total
0      Afganistán                AFG  1750  0.000000e+00
1      Afganistán                AFG  1751  0.000000e+00
2      Afganistán                AFG  1752  0.000000e+00
3      Afganistán                AFG  1753  0.000000e+00
4      Afganistán                AFG  1754  0.000000e+00
...           ...                ...   ...           ...
63099      Global                WLD  2017  3.609674e+10
63100      Global                WLD  2018  3.682651e+10
63101      Global                WLD  2019  3.708256e+10
63102      Global                WLD  2020  3.526409e+10
63103      Global                WLD  2021  3.712385e+10

[63104 rows x 4 columns]

#### **Google Sheets**

Pasos:

1. Subir el archivo a Google Sheets
2. Compartir el archivo, asegurando de dar acceso a cualquier persona
3. Copiar el enlace
https://docs.google.com/spreadsheets/d/1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ/edit?usp=sharing

4. Extraer el ID: Se encuentra desde D/ hasta antes de la siguiente /
1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ

5. Creamos las variables para la ID y el enlace

* Al enlace (url) le eliminamos lo que viene luego del / de la ID, y lo
reemplazamos por `/gviz/tq`  -> que permite la comunicación entre Google Sheets y Google Colab (o cualquier otra herramienta que se esté utilizando para acceder a los datos.

* Siguiendo con el enlace (url), le adicionamos `tqx=out:csv` -> que especifíca el formato en el que se desean recibir los datos desde Google Sheets.
`En este caso, out=csv indica que quieres que los datos se exporten en formato CSV`.

* Y finalmente en nuestra url le agregamos `&sheet` -> que identifica que hoja de cálculo dentro del archivo de Google Sheets se quiere leer.

Quendando algo como esto:

```
sheet_id = '1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ'
url_sheet = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'
```



In [55]:
# Identificamos y modificamos nuestra url
sheet_id = '1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ'
url_sheet = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'

In [56]:
# Lectura del archivo y almacenarlo en un DataFrame
datos_co2_sheets = pd.read_csv(url_sheet)

In [57]:
# Mostrar el DataFrame
datos_co2_sheets

País ISO 3166-1 alpha-3   Año           Total          Carbón  \
0      Afganistán                AFG  1750               0             NaN   
1      Afganistán                AFG  1751               0             NaN   
2      Afganistán                AFG  1752               0             NaN   
3      Afganistán                AFG  1753               0             NaN   
4      Afganistán                AFG  1754               0             NaN   
...           ...                ...   ...             ...             ...   
63099      Global                WLD  2017  36,096,739,276  14,506,973,805   
63100      Global                WLD  2018  36,826,506,600  14,746,830,688   
63101      Global                WLD  2019  37,082,558,969  14,725,978,025   
63102      Global                WLD  2020  35,264,085,734  14,174,564,010   
63103      Global                WLD  2021  37,123,850,352  14,979,598,083   

               Aceite            Gas        Cemento        Quema        Otros  \
0                 NaN            NaN            NaN          NaN          NaN   
1                 NaN            NaN            NaN          NaN          NaN   
2                 NaN            NaN            NaN          NaN          NaN   
3                 NaN            NaN            NaN          NaN          NaN   
4                 NaN            NaN            NaN          NaN          NaN   
...               ...            ...            ...          ...          ...   
63099  12,242,627,935  7,144,928,128  1,507,923,185  391,992,176  302,294,047   
63100  12,266,016,285  7,529,846,784  1,569,218,392  412,115,746  302,478,706   
63101  12,345,653,374  7,647,528,220  1,617,506,786  439,253,991  306,638,573   
63102  11,191,808,551  7,556,290,283  1,637,537,532  407,583,673  296,301,685   
63103  11,837,159,116  7,921,829,472  1,672,592,372  416,525,563  296,145,746   

      Per Capita  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
63099  4,749,682  
63100  4,792,753  
63101  4,775,633  
63102  4,497,423  
63103  4,693,699  

[63104 rows x 11 columns]

In [58]:
# Identificamos y modificamos nuestra url según nuestra hoja de cálculo
sheet_id = '1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ'
sheet_name_2 = 'emisiones_percapita'
url_percapita = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name_2}'

In [59]:
# Lectura del archivo y almacenarlo en un DataFrame
datos_percapita_sheets = pd.read_csv(url_percapita)

In [60]:
# Mostrar el DataFrame
datos_percapita_sheets

País ISO 3166-1 alpha-3   Año      Total     Carbón     Aceite  \
0      Afganistán                AFG  1750        NaN        NaN        NaN   
1      Afganistán                AFG  1751        NaN        NaN        NaN   
2      Afganistán                AFG  1752        NaN        NaN        NaN   
3      Afganistán                AFG  1753        NaN        NaN        NaN   
4      Afganistán                AFG  1754        NaN        NaN        NaN   
...           ...                ...   ...        ...        ...        ...   
63099      Global                WLD  2017  4,749,682  1,908,857  1,610,910   
63100      Global                WLD  2018  4,792,753  1,919,213  1,596,350   
63101      Global                WLD  2019  4,775,633  1,896,468  1,589,920   
63102      Global                WLD  2020  4,497,423  1,807,760  1,427,353   
63103      Global                WLD  2021  4,693,699  1,893,923  1,496,614   

             Gas  Cemento   Quema   Otros  
0            NaN      NaN     NaN     NaN  
1            NaN      NaN     NaN     NaN  
2            NaN      NaN     NaN     NaN  
3            NaN      NaN     NaN     NaN  
4            NaN      NaN     NaN     NaN  
...          ...      ...     ...     ...  
63099    940,144  198,416  51,579  39,776  
63100    979,965  204,225  53,634  39,366  
63101    984,878  208,309  56,569  39,490  
63102    963,695  208,844  51,981  37,789  
63103  1,001,585  211,472  52,663  37,443  

[63104 rows x 10 columns]

In [61]:
# Identificamos y modificamos nuestra url según nuestra hoja de cálculo
sheet_id = '1MisXtwDT5Y6p1zYokhrpYiKOUAXMiuP8AR9f2sgjbHQ'
sheet_name_3 = 'fuentes'
url_fuentes = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name_3}'

In [62]:
# Lectura del archivo y almacenarlo en un DataFrame
datos_fuentes_sheets = pd.read_csv(url_fuentes)

In [63]:
# Mostrar el DataFrame
datos_fuentes_sheets.tail()

País ISO 3166-1 alpha-3   Año                                 Total  \
63099  Global                WLD  2017  CDIAC 2022, BP, and Sum of countries   
63100  Global                WLD  2018  CDIAC 2022, BP, and Sum of countries   
63101  Global                WLD  2019  CDIAC 2022, BP, and Sum of countries   
63102  Global                WLD  2020  CDIAC 2022, BP, and Sum of countries   
63103  Global                WLD  2021  CDIAC 2022, BP, and Sum of countries   

                  Carbón             Aceite                Gas        Cemento  \
63099  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63100  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63101  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63102  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   
63103  CDIAC 2022 and BP  CDIAC 2022 and BP  CDIAC 2022 and BP  Andrew cement   

                    Quema   Otros  \
63099  CDIAC 2022 and GCP  [NONE]   
63100  CDIAC 2022 and GCP  [NONE]   
63101  CDIAC 2022 and GCP  [NONE]   
63102  CDIAC 2022 and GCP  [NONE]   
63103  CDIAC 2022 and GCP  [NONE]   

                                              Per Capita  
63099  CDIAC 2022, BP, Sum of countries, and UN popul...  
63100  CDIAC 2022, BP, Sum of countries, and UN popul...  
63101  CDIAC 2022, BP, Sum of countries, and UN popul...  
63102  CDIAC 2022, BP, Sum of countries, and UN popul...  
63103  CDIAC 2022, BP, Sum of countries, and UN popul...

### Desafío: leyendo datos de otro link

#### Ejercicio:

* Es hora de que pongas a prueba los conocimientos desarrollados durante la clase. Tenemos un [link de Google Sheets](https://cdn3.gnarususercontent.com.br/1980-pandas-e-s/emisiones_carbono_mundo.xlsx
) que contiene datos importantes sobre las emisiones de dióxido de carbono en todo el mundo. El conjunto de datos se obtuvo de [Kaggle](https://www.kaggle.com/datasets/koustavghosh149/co2-emission-around-the-world) y consta de las emisiones de CO2 per cápita de todos los países del mundo entre 1990 y 2019.


* En este desafío, tu función es leer este link de Google Sheets y luego guardar el DataFrame obtenido en formato CSV.

In [64]:
# Identificamos y modificamos nuestra url según nuestra hoja de cálculo
sheet_id_ejercicio_2 = '1O9eZ5IrbOVzkJN8XIoGru97afGIZyYwmq47JQsjRh0w'
url_ejercicio_2 = f'https://docs.google.com/spreadsheets/d/{sheet_id_ejercicio_2}/gviz/tq?tqx=out:csv&sheet'

In [65]:
# Lectura del archivo y almacenarlo en un DataFrame
datos_ejercicio_2 = pd.read_csv(url_ejercicio_2)

In [66]:
# Mostrar el DataFrame
datos_ejercicio_2.head()

Country Name country_code                     Region  \
0        Aruba          ABW  Latin America & Caribbean   
1  Afghanistan          AFG                 South Asia   
2       Angola          AGO         Sub-Saharan Africa   
3      Albania          ALB      Europe & Central Asia   
4      Andorra          AND      Europe & Central Asia   

                           Indicator Name           1990         1991  \
0  CO2 emissions (metric tons per capita)            NaN          NaN   
1  CO2 emissions (metric tons per capita)    191,745,115  167,681,579   
2  CO2 emissions (metric tons per capita)    553,661,955  544,538,649   
3  CO2 emissions (metric tons per capita)  1,819,541,634  124,281,022   
4  CO2 emissions (metric tons per capita)  7,521,831,658  723,537,924   

          1992           1993           1994           1995  ...  \
0          NaN            NaN            NaN            NaN  ...   
1   95,957,742      8,472,111     75,545,827     68,467,956  ...   
2  543,557,223     70,898,423    836,804,405    912,141,485  ...   
3  683,699,826    638,307,036    645,355,189    605,436,253  ...   
4   69,630,787  6,724,177,518  6,541,578,914  6,733,479,486  ...   

            2011           2012           2013           2014           2015  \
0            NaN            NaN            NaN            NaN            NaN   
1     29,650,624    259,295,334     18,562,366    146,235,617    172,896,741   
2    985,522,312    950,695,879  1,036,293,852  1,099,779,111    113,504,405   
3    166,942,319    150,324,046  1,533,630,039  1,668,337,371    160,377,515   
4  5,850,886,105  5,944,654,173  5,942,800,412  5,807,127,723  6,026,181,822   

            2016           2017           2018           2019         2019.1  
0            NaN            NaN            NaN            NaN            NaN  
1    149,789,334    131,694,556      1,632,953    159,824,372    159,824,372  
2  1,031,811,348     81,330,073    777,674,934    792,137,069    792,137,069  
3  1,557,664,358  1,788,786,074  1,782,738,948    169,224,832    169,224,832  
4  6,080,600,282  6,104,133,912  6,362,975,399  6,481,217,432  6,481,217,432  

[5 rows x 35 columns]

In [67]:
# Guarda el DataFrame en un archivo CSV
datos_ejercicio_2.to_csv('/content/drive/MyDrive/Pandas_2/archivos_excel/emisiones_carbono_mundial.csv', index=False)

In [68]:
# Leemos el archivo CSV creado
pd.read_csv('/content/drive/MyDrive/Pandas_2/archivos_excel/emisiones_carbono_mundial.csv').head()

Country Name country_code                     Region  \
0        Aruba          ABW  Latin America & Caribbean   
1  Afghanistan          AFG                 South Asia   
2       Angola          AGO         Sub-Saharan Africa   
3      Albania          ALB      Europe & Central Asia   
4      Andorra          AND      Europe & Central Asia   

                           Indicator Name           1990         1991  \
0  CO2 emissions (metric tons per capita)            NaN          NaN   
1  CO2 emissions (metric tons per capita)    191,745,115  167,681,579   
2  CO2 emissions (metric tons per capita)    553,661,955  544,538,649   
3  CO2 emissions (metric tons per capita)  1,819,541,634  124,281,022   
4  CO2 emissions (metric tons per capita)  7,521,831,658  723,537,924   

          1992           1993           1994           1995  ...  \
0          NaN            NaN            NaN            NaN  ...   
1   95,957,742      8,472,111     75,545,827     68,467,956  ...   
2  543,557,223     70,898,423    836,804,405    912,141,485  ...   
3  683,699,826    638,307,036    645,355,189    605,436,253  ...   
4   69,630,787  6,724,177,518  6,541,578,914  6,733,479,486  ...   

            2011           2012           2013           2014           2015  \
0            NaN            NaN            NaN            NaN            NaN   
1     29,650,624    259,295,334     18,562,366    146,235,617    172,896,741   
2    985,522,312    950,695,879  1,036,293,852  1,099,779,111    113,504,405   
3    166,942,319    150,324,046  1,533,630,039  1,668,337,371    160,377,515   
4  5,850,886,105  5,944,654,173  5,942,800,412  5,807,127,723  6,026,181,822   

            2016           2017           2018           2019         2019.1  
0            NaN            NaN            NaN            NaN            NaN  
1    149,789,334    131,694,556      1,632,953    159,824,372    159,824,372  
2  1,031,811,348     81,330,073    777,674,934    792,137,069    792,137,069  
3  1,557,664,358  1,788,786,074  1,782,738,948    169,224,832    169,224,832  
4  6,080,600,282  6,104,133,912  6,362,975,399  6,481,217,432  6,481,217,432  

[5 rows x 35 columns]

## **Aula 3: Leyendo archivos JSON**

JSON: JavaScript Object Notation

* Es un formato de texto ligero y fácil de leer/escribir que sirve para almacenar y transferir datos estructurados, especialmente entre servidores y aplicaciones web.

Un archivo JSON está compuesto por:

* Llaves (keys): Son los nombres de los campos.
* Valores (values): Son los datos asociados a esas llaves.

Para visualizar un archivo JSON, iremos a la siguiente página:
https://jsonviewer.stack.hu/


Para visualizar un archivo JSON con información anidada utilizaremos: https://jsoncrack.com/editor > Import

### **Importando archivos JSON**

In [69]:
import pandas as pd

#### **Lectura de archivo JSON (Sin información anidada)**

In [70]:
# Lectura del archivo JSON sin información anidada
datos_pacientes = pd.read_json('/content/drive/MyDrive/Pandas_2/archivos_json/pacientes.json')

In [71]:
# Mostrar el DataFrame
datos_pacientes

ID_paciente Enfermedad_corazon    IMC Fumador Consumo_alcohol AVC  \
0              0                 No  16.60      Sí              No  No   
1              1                 No  20.34      No              No  Sí   
2              2                 No  26.58      Sí              No  No   
3              3                 No  24.21      No              No  No   
4              4                 No  23.71      No              No  No   
..           ...                ...    ...     ...             ...  ..   
995          995                 No  28.29      Sí              No  No   
996          996                 No  25.69      Sí              No  No   
997          997                 Sí  26.99      No              No  No   
998          998                 No  44.29      No              No  No   
999          999                 No  25.83      No              No  No   

     Salud_fisica  Salud_mental Dificultad_caminar Sexo_biologico Rango_edad  \
0               3            30                 No       Femenino      55-59   
1               0             0                 No       Femenino     80 ó +   
2              20            30                 No      Masculino      65-69   
3               0             0                 No       Femenino      75-79   
4              28             0                 Sí       Femenino      40-44   
..            ...           ...                ...            ...        ...   
995            10            20                 No       Femenino      65-69   
996             0             0                 No       Femenino      75-79   
997             1             0                 No      Masculino      70-74   
998            30             0                 Sí       Femenino      70-74   
999             0             0                 No      Masculino      70-74   

       Raza Diabetes Actividad_fisica Salud_general  Horas_sueño Asma  \
0    Blanca       Sí               Sí     Muy buena            5   Sí   
1    Blanca       No               Sí     Muy buena            7   No   
2    Blanca       Sí               Sí     Razonable            8   Sí   
3    Blanca       No               No         Buena            6   No   
4    Blanca       No               Sí     Muy buena            8   No   
..      ...      ...              ...           ...          ...  ...   
995   Negra       Sí               Sí     Razonable            8   No   
996  Blanca       No               Sí     Excelente            8   No   
997  Blanca       No               Sí     Muy buena            8   No   
998  Blanca       Sí               No         Buena            6   Sí   
999   Negra       Sí               No     Excelente            7   No   

    Enfermedad_renal Cancer_piel  
0                 No          Sí  
1                 No          No  
2                 No          No  
3                 No          Sí  
4                 No          No  
..               ...         ...  
995               No          No  
996               No          No  
997               No          No  
998               No          No  
999               No          No  

[1000 rows x 19 columns]

#### **Lectura de archivo JSON (Con información anidada)**

In [72]:
# Lectura del archivo JSON sin información anidada
datos_pacientes_2 = pd.read_json('/content/drive/MyDrive/Pandas_2/archivos_json/pacientes_2.json')

In [73]:
# Mostrar el DataFrame
datos_pacientes_2

# Vemos que el DataFrame nos entrega información errónea,
# ya que hay un diccionario dentro de otro diccionario.

# A esto de le llama tener información anidada.
# Para esto debemos NORMALIZAR la información

Investigación   Año  \
0  Indicadores clave de enfermedades cardíacas  2020   
1  Indicadores clave de enfermedades cardíacas  2020   
2  Indicadores clave de enfermedades cardíacas  2020   

                                           Pacientes  
0  {'ID': '01', 'Rango_edad': '55-59', 'Sexo_biol...  
1  {'ID': '02', 'Rango_edad': '80 ó +', 'Sexo_bio...  
2  {'ID': '03', 'Rango_edad': '65-69', 'Sexo_biol...

Para visualizar un archivo JSON con información anidada utilizaremos: https://jsoncrack.com/editor > Import

#### **Normalizando archivos JSON**

Se debe normalizar los archivos JSON cuando encontramos información anidada.

In [74]:
# Debemos indicar en que columna tenemos información anidada

# En este caso, nos dimos cuenta que el archivo maneja información de 3 pacientes
# y que el campo/columna Pacientes tiene información anidada

# Normalizamos la información y la almacenamos en un DataFrame
df_normalizado = pd.json_normalize(datos_pacientes_2['Pacientes'])

In [75]:
# Mostramos el DataFrame
df_normalizado

ID Rango_edad Sexo_biologico    Raza    IMC Fumador Consumo_alcohol  \
0  01      55-59          Mujer  Blanca  16.60      Sí              No   
1  02     80 ó +          Mujer  Blanca  20.34      No              No   
2  03      65-69      Masculino  Blanca  26.58      Sí              No   

   Salud_física  Salud_mental Dificultad_caminar Actividad_física  \
0             3            30                 No               Sí   
1             0             0                 No               Sí   
2            20            30                 No               Sí   

  Salud_general  Horas_sueño                Problemas_salud  
0     Muy buena            5  [Diabetes, Asma, Cancer_piel]  
1     Muy buena            7                          [AVC]  
2     Muy buena            8               [Diabetes, Asma]

### **Exportando archivo JSON**

In [76]:
# Exportamos el DataFrame a un archivo JSON
df_normalizado.to_json('/content/drive/MyDrive/Pandas_2/archivos_json/historico_pacientes_normalizado.json')

In [77]:
# Leemos el archivo exportado
pd.read_json('/content/drive/MyDrive/Pandas_2/archivos_json/historico_pacientes_normalizado.json')

ID Rango_edad Sexo_biologico    Raza    IMC Fumador Consumo_alcohol  \
0   1      55-59          Mujer  Blanca  16.60      Sí              No   
1   2     80 ó +          Mujer  Blanca  20.34      No              No   
2   3      65-69      Masculino  Blanca  26.58      Sí              No   

   Salud_física  Salud_mental Dificultad_caminar Actividad_física  \
0             3            30                 No               Sí   
1             0             0                 No               Sí   
2            20            30                 No               Sí   

  Salud_general  Horas_sueño                Problemas_salud  
0     Muy buena            5  [Diabetes, Asma, Cancer_piel]  
1     Muy buena            7                          [AVC]  
2     Muy buena            8               [Diabetes, Asma]

### **Desafío: normalizar el JSON obtenido de la API**

Es el momento de poner a prueba tus conocimientos adquiridos durante la clase. El DataFrame que se muestra a continuación se generó después de obtener los datos de la API de Fruitvice:


| index | genus    | name       | id | family    | order        | nutritions                                                                 |
|-------|----------|------------|----|-----------|--------------|----------------------------------------------------------------------------|
| 0     | Malus    | Apple      | 6  | Rosaceae  | Rosales      | {'carbohydrates': 11.4, 'protein': 0.3, 'fat': 0.4, 'calories': 52, 'sugar': 10.3} |
| 1     | Prunus   | Apricot    | 35 | Rosaceae  | Rosales      | {'carbohydrates': 3.9, 'protein': 0.5, 'fat': 0.1, 'calories': 15, 'sugar': 3.2}  |
| 2     | Persea   | Avocado    | 84 | Lauraceae | Laurales     | {'carbohydrates': 8.53, 'protein': 2, 'fat': 14.66, 'calories': 160, 'sugar': 0.66} |
| 3     | Musa     | Banana     | 1  | Musaceae  | Zingiberales | {'carbohydrates': 22, 'protein': 1, 'fat': 0.2, 'calories': 96, 'sugar': 17.2}  |
| 4     | Rubus    | Blackberry | 64 | Rosaceae  | Rosales      | {'carbohydrates': 9, 'protein': 1.3, 'fat': 0.4, 'calories': 40, 'sugar': 4.5}   |
| 5     | Fragaria | Blueberry  | 33 | Rosaceae  | Rosales      | {'carbohydrates': 5.5, 'protein': 0, 'fat': 0.4, 'calories': 29, 'sugar': 5.4}   |
| 6     | Prunus   | Cherry     | 9  | Rosaceae  | Rosales      | {'carbohydrates': 12, 'protein': 1, 'fat': 0.3, 'calories': 50, 'sugar': 8}     |



* Este DataFrame tiene 6 columnas: genus (género), name (nombre), id, family (familia), order (orden) e nutritions (nutriciones). Observe que la columna nutritions tiene toda la información nutricional anidada en ella. Por tanto, es necesario normalizar los datos.

El desafío ahora es normalizar este DataFrame.

In [78]:
# Importa la biblioteca requests, que se utiliza para hacer solicitudes HTTP a servidores web.
import requests
# Importa el módulo json, que permite trabajar con datos en formato JSON (codificar y decodificar).
import json

In [79]:
# Utiliza la función get de la biblioteca requests para hacer una solicitud a la API Fruitvice y obtener los datos de todas las frutas en formato JSON.
# La respuesta del servidor se guarda en la variable datos_frutas.
respuesta_api = requests.get('https://fruityvice.com/api/fruit/all')

In [80]:
#  Transforma el texto JSON recibido en datos que Python puede entender.
resultado = json.loads(respuesta_api.text)

In [81]:
# Organiza los datos en una tabla (DataFrame) para analizarlos fácilmente.
pd.DataFrame(resultado).head()

name  id      family         order      genus  \
0   Persimmon  52   Ebenaceae       Rosales  Diospyros   
1  Strawberry   3    Rosaceae       Rosales   Fragaria   
2      Banana   1    Musaceae  Zingiberales       Musa   
3      Tomato   5  Solanaceae     Solanales    Solanum   
4        Pear   4    Rosaceae       Rosales      Pyrus   

                                          nutritions  
0  {'calories': 81, 'fat': 0.0, 'sugar': 18.0, 'c...  
1  {'calories': 29, 'fat': 0.4, 'sugar': 5.4, 'ca...  
2  {'calories': 96, 'fat': 0.2, 'sugar': 17.2, 'c...  
3  {'calories': 74, 'fat': 0.2, 'sugar': 2.6, 'ca...  
4  {'calories': 57, 'fat': 0.1, 'sugar': 10.0, 'c...

#### **Ejercicio:**

In [82]:
import pandas as pd
import requests
import json

# Obtener la respuesta de la Api
respuesta_api = requests.get('https://fruityvice.com/api/fruit/all')

In [83]:
# Convertir el contenido JSON a estructura Python
datos_frutas = respuesta_api.json()

In [84]:
# Normalizamos la información y la almacenamos en un DataFrame
frutas_normalizado = pd.json_normalize(datos_frutas)

In [85]:
# Mostramos el DataFrame
frutas_normalizado.head()

name  id      family         order      genus  nutritions.calories  \
0   Persimmon  52   Ebenaceae       Rosales  Diospyros                   81   
1  Strawberry   3    Rosaceae       Rosales   Fragaria                   29   
2      Banana   1    Musaceae  Zingiberales       Musa                   96   
3      Tomato   5  Solanaceae     Solanales    Solanum                   74   
4        Pear   4    Rosaceae       Rosales      Pyrus                   57   

   nutritions.fat  nutritions.sugar  nutritions.carbohydrates  \
0             0.0              18.0                      18.0   
1             0.4               5.4                       5.5   
2             0.2              17.2                      22.0   
3             0.2               2.6                       3.9   
4             0.1              10.0                      15.0   

   nutritions.protein  
0                 0.0  
1                 0.8  
2                 1.0  
3                 0.9  
4                 0.4

In [86]:
# Exportamos el DataFrame a un archivo JSON normalizado
frutas_normalizado.to_json('/content/drive/MyDrive/Pandas_2/archivos_json/frutas.json')

In [87]:
# Lectura del archivo JSON creado
pd.read_json('/content/drive/MyDrive/Pandas_2/archivos_json/frutas.json').head()

name  id      family         order      genus  nutritions.calories  \
0   Persimmon  52   Ebenaceae       Rosales  Diospyros                   81   
1  Strawberry   3    Rosaceae       Rosales   Fragaria                   29   
2      Banana   1    Musaceae  Zingiberales       Musa                   96   
3      Tomato   5  Solanaceae     Solanales    Solanum                   74   
4        Pear   4    Rosaceae       Rosales      Pyrus                   57   

   nutritions.fat  nutritions.sugar  nutritions.carbohydrates  \
0             0.0              18.0                      18.0   
1             0.4               5.4                       5.5   
2             0.2              17.2                      22.0   
3             0.2               2.6                       3.9   
4             0.1              10.0                      15.0   

   nutritions.protein  
0                 0.0  
1                 0.8  
2                 1.0  
3                 0.9  
4                 0.4